plugins/org.eclipse.glsp.server/src/org/eclipse/glsp/server/internal/actions/DefaultActionDispatcher.java

```diff
@@ -21,80 +21,56 @@
 import java.util.stream.Collectors;
 
 import org.apache.log4j.Logger;
 import org.eclipse.glsp.server.actions.Action;
 import org.eclipse.glsp.server.actions.ActionDispatcher;
+import org.eclipse.glsp.server.actions.ActionExecutor;
 import org.eclipse.glsp.server.actions.ActionHandler;
 import org.eclipse.glsp.server.actions.ActionHandlerRegistry;
 import org.eclipse.glsp.server.actions.ResponseAction;
 import org.eclipse.glsp.server.di.ClientId;
-import org.eclipse.glsp.server.disposable.Disposable;
 import org.eclipse.glsp.server.features.core.model.UpdateModelAction;
-import org.eclipse.glsp.server.protocol.GLSPClient;
 import org.eclipse.glsp.server.utils.FutureUtil;
 
 import com.google.inject.Inject;
-import com.google.inject.Provider;
 
 /**
  * <p>
- * An ActionDispatcher that executes all handlers in the same thread. The dispatcher's
+ * An ActionDispatcher that executes all handlers with the {@link ActionExecutor}. The dispatcher's
  * public methods can be invoked from any thread, e.g. from background jobs running on
  * the server.
  * </p>
  */
-public class DefaultActionDispatcher extends Disposable implements ActionDispatcher {
+public class DefaultActionDispatcher implements ActionDispatcher {
 
    private static final Logger LOG = Logger.getLogger(DefaultActionDispatcher.class);
 
-   private static final AtomicInteger COUNT = new AtomicInteger(0);
-
    @Inject
    protected ActionHandlerRegistry actionHandlerRegistry;
 
+   @Inject
+   protected ActionExecutor actionExecutor;
+
    @Inject
    @ClientId
    protected String clientId;
 
-   protected final String name;
-
-   protected final Thread thread;
-
-   protected final BlockingQueue<Action> actionsQueue = new ArrayBlockingQueue<>(100, true);
-
    protected List<Action> postUpdateQueue = new ArrayList<>();
 
    // Results will be placed in the map when the action dispatcher receives a new action (From arbitrary threads),
    // and will be removed from the dispatcher's thread.
    protected final Map<Action, CompletableFuture<Void>> results = Collections.synchronizedMap(new HashMap<>());
 
-   // Use a provider, as the GLSPClient is probably not created yet. We won't receive
-   // any action until it's ready anyway.
-   @Inject
-   protected Provider<GLSPClient> client;
-
-   public DefaultActionDispatcher() {
-      this.name = getClass().getSimpleName() + " " + COUNT.incrementAndGet();
-      this.thread = new Thread(this::runThread);
-      this.thread.setName(this.name);
-      this.thread.setDaemon(true);
-      this.thread.start();
-   }
-
    @Override
    public CompletableFuture<Void> dispatch(final Action action) {
-
       CompletableFuture<Void> result = new CompletableFuture<>();
       results.put(action, result);
-      if (thread == Thread.currentThread()) {
-         // Actions dispatched from the ActionDispatcher thread don't have to go back
+      if (actionExecutor.isExecutorThread(Thread.currentThread())) {
```

Comment:
```
public CompletableFuture<Void> dispatch(final Action action) { 
    return dispatch(action, false); 
}
```

This assumes that this method will always be invoked from a separate thread. However, I think this would be a risky assumption. Actions may return sub-actions, but they could also trigger a service that actually needs to immediately execute an Action, and would call the ActionDispatcher for that. Such calls need to be intercepted as well.

So I think it's a lot safer to keep the isExecutionThread(). We *really* don't want to call addToQueue() from the execution thread, as it would cause a guaranteed deadlock.